### PR TITLE
feat: specialized fwd/bwd cutedsl kernel for d512

### PR DESCRIPTION
**Pull request by Butterfingrz**
# feat(cutedsl): add SM90 specialized attention kernel for head_dim 512

## 📝 Summary

新增 `ffpa_attn.cutedsl` 后端，提供专门针对 **SM90 (Hopper) + `head_dim == 512`** 优化的前向与反向 attention kernel

本次更新通过两个公开入口将新内核接入现有的 FFPA 统一接口：
- **`ffpa_attn.ffpa_attn_func`**：支持固定形状 `[B, Nh, N, D]`。通过显式声明 `forward_backend='cutedsl'` 开启（opt-in）；默认仍回退至 Triton。
- **`ffpa_attn.ffpa_attn_varlen_func`**：**【新增】** 类似 FlashAttention 风格的 packed THD 变长 API。`cutedsl` 是该接口的唯一后端。

*注：所有路径最终都会收敛到同一对底层函数 `_flash_attn_fwd_sm90` / `_flash_attn_bwd_sm90`，保证数值位级一致性。*

---

### 1. 核心后端包 (`src/ffpa_attn/cutedsl/`)
完整引入了 CuTeDSL 后端，包含以下核心模块：
- **`_ffpa_fwd_d512_sm90.py`**：前向 full-D 三 Warpgroup 流水线（TMA producer + WG1 QK/softmax/PV-front + WG2 PV-back epilogue）。
- **`_ffpa_dkdv_d512_sm90.py`**：反向 dK/dV。1 TMA + 1 MMA WG 架构，`d_chunk=256`，K/V 在 SMEM 中跨 d-pass 复用。
- **`_ffpa_dq_d512_sm90.py`**：反向 dQ。双非对称 MMA WG 架构，dQ_front + dQ_back 协作。
- **`flash_bwd_preprocess.py`**：`D_i = (O ⊙ dO).sum(-1)` 预处理，返回带有 seq padding 的 LSE
- **`interface.py`**：包含 `torch.library` ops (`splitd_flash_attn::varlen_{fwd,bwd}`)、`FlashAttnFunc` autograd、`split_flash_attn_func` 快路径入口，以及针对 SM90 / dtype / GQA-pack 的严格守卫（Guards）。
- **`_wrappers.py`**：负责 SDPA 布局 `[B, H, N, D]` 与 CuTeDSL 原生布局 `[B, N, H, D]` 的互相转置，并充当唯一的能力闸门（`_require_cutedsl_supported`）。
- **`utils/`**：底层工具链，包含：tile scheduler, named barrier, online softmax, mask 处理 (causal/local/mask_mod), SeqlenInfoQK, BlockInfo, pack_gqa, PipelineAsync/PipelineTmaAsync, JIT/AOT cache, ptxas hook, FA_LOG_LEVEL 日志与 TVM FFI 流集成。

### 2. 接口与路由集成
- **`ffpa_attn_interface.py`**：
  - 新增 `_should_take_cutedsl_fast_path` 与 `_ffpa_attn_func_cutedsl_fast_path`。
  - **Fast Path**：当命中 4 个特定条件时，直接调用 `split_flash_attn_func`。
  - **Slow Path**：未命中时回退到标准慢路径，保持错误信息行为一致。
- **`functional.py`**：
  - `FFPAAttnMeta` 现已支持 `forward_backend='cutedsl'` / `backward_backend='cutedsl'`（前/反向必须成对绑定）。
  - 在 `normalize_inputs` 阶段会主动调用 `_require_cutedsl_supported`。如果遇到不支持的场景（如 $D \neq 512$ / 非 SM90 / fp16 训练 / dropout / 存在 attn_mask），将直接抛出 **actionable error**，而不会进行静默 fallback。

### 3. 基准测试 & 正确性验证
- **Benchmarks**：新增 H200 上的 $D=512$ 加速曲线报告（详见 `docs/assets/perf/ffpa_speedup_cutedsl_*`）。

### 🚀 Performance 

我们在 **NVIDIA H200** 上以 PyTorch 官方的 `SDPA` 作为 Baseline，测试了四种典型 workload 的吞吐量与加速比（前向与反向均参与计算）。

### Throughput (TFLOPS)

| Workload | SDPA FWD | FFPA FWD | SDPA BWD | FFPA BWD |
| :--- | :---: | :---: | :---: | :---: |
| **self-attn** | 134 | **385** | 45 | **234** |
| **cross-attn** | 135 | **257** | 46 | **215** |
| **gqa** | 134 | **395** | 45 | **243** |
| **causal** | 131 | **346** | 42 | **227** |

### Speedup vs SDPA

| Workload | FFPA FWD | FFPA BWD |
| :--- | :---: | :---: |
| **self-attn** | **2.9x** | **5.2x** |
| **cross-attn** | **1.9x** | **4.7x** |
| **gqa** | **2.9x** | **5.4x** |
| **causal** | **2.6x** | **5.5x** |

> **Note**: 测试结果显示，在反向传播 (BWD) 阶段取得了尤为显著的加速（最高达 5.5x），前向 (FWD) 阶段仍可以继续优化

- **Tests**：
  - `tests/test_ffpa_cutedsl.py`：校验 fast-path 与 autograd，确保与 Triton 实现的结果对齐。

### 4. 依赖打包
- **`pyproject.toml`**：新增 `cutedsl` extra 依赖组（包含 `quack-kernels`、`nvidia-cutlass-dsl`、`apache-tvm-ffi`）。

### CODE REVIEW

File: examples/perf_cutedsl.py [hunk 1]
```diff
@@ -0,0 +1,1105 @@
+"""Benchmark the CuTeDSL D=512 SM90 FFPA path and render plot plus Markdown tables.
```
Comment (DefTruth) [inline examples/perf_cutedsl.py]: 合并到perf.py中，通过--forward-backend cutedsl, --backwward-backend cutedsl 来走cutedsl分支

File: src/ffpa_attn/cutedsl/_ffpa_bwd_preprocess.py [hunk 2]
```diff
@@ -0,0 +1,332 @@
+# This file is copied from https://github.com/Dao-AILab/flash-attention/blob/main/flash_attn/cute/flash_bwd_preprocess.py
```
Comment (DefTruth) [inline src/ffpa_attn/cutedsl/_ffpa_bwd_preprocess.py:1]: rename flash_bwd_preprocess.py -> _bwd_preprocess.py

File: src/ffpa_attn/cutedsl/README.md [hunk 3]
```diff
@@ -0,0 +1,246 @@
+# FFPA CuTeDSL Backend (`ffpa_attn.cutedsl`)
```
Comment (DefTruth) [inline src/ffpa_attn/cutedsl/README.md:1]: interface.py -> _interface.py

File: src/ffpa_attn/ffpa_attn_interface.py [hunk 4]
```diff
@@ -170,6 +271,21 @@ def ffpa_attn_func(
       enable_gqa=enable_gqa,
     )
 
+  # D=512 SM90 CuTeDSL fast-path — opt-in via forward_backend='cutedsl'.
+  # Bypasses FFPAAttnMeta normalization and the FFPAAttnFunc autograd boundary,
+  # dispatching directly to cutedsl.interface.split_flash_attn_func (autograd
+  # via FlashAttnFunc). Ineligible cutedsl requests fall through so the slow
+  # path raises the canonical NotImplementedError.
+  if _should_take_cutedsl_fast_path(query, attn_mask, dropout_p, meta.forward_backend):
```
Comment (DefTruth) [inline src/ffpa_attn/ffpa_attn_interface.py]: 关于backend的选择，当前希望是用户明确指定cutedsl后端才走cutedsl的实现。暂时不走_should_take_cutedsl_fast_path这种自动切换的逻辑。未来可以考虑增加一个auto模式来支持。

File: src/ffpa_attn/ffpa_attn_interface.py [hunk 5]
```diff
@@ -184,4 +300,215 @@ def ffpa_attn_func(
   return FFPAAttnFunc.apply(query, key, value, attn_bias, meta)
 
 
-__all__ = ["ffpa_attn_func"]
+def _ffpa_attn_varlen_cutedsl(
```
Comment (DefTruth) [inline src/ffpa_attn/ffpa_attn_interface.py]: 这些 _xxx_cutedsl的API，都是内部函数，不应该在ffpa_attn_interface.py中实现，应该方向cutedsl/目录下实现，在ffpa_attn_interface.py只做调用。

File: src/ffpa_attn/cutedsl/_ffpa_fwd_d512_sm90.py [hunk 6]
```diff
@@ -0,0 +1,1716 @@
...
+from .utils.tile_scheduler import (
+  TileSchedulerArguments,
+  SingleTileScheduler,
+  SingleTileLPTScheduler,
+  SingleTileVarlenScheduler,
+)
+from cutlass.cute import FastDivmodDivisor
+
+
+class FlashAttentionForwardSm90TrainOnly:
```
Comment (DefTruth) [inline src/ffpa_attn/cutedsl/_ffpa_fwd_d512_sm90.py]: 我们应该维护一套干净的FFPA namespace/classname/funcname便于后期维护和扩展，而不是和flash混用。从flash从copy/adapted过来的，明确注明来源。其他namespace/classname/funcname混用的地方按照这个逻辑一起修改。

**Review by DefTruth — CHANGES_REQUESTED** [review]

File: examples/perf_cutedsl.py (hunk 1 above)
Comment (Butterfingrz) [reply examples/perf_cutedsl.py]: 已合并

File: src/ffpa_attn/cutedsl/_ffpa_bwd_preprocess.py (hunk 2 above)
Comment (Butterfingrz) [reply src/ffpa_attn/cutedsl/_ffpa_bwd_preprocess.py:1]: 已修改

File: src/ffpa_attn/cutedsl/README.md (hunk 3 above)
Comment (Butterfingrz) [reply src/ffpa_attn/cutedsl/README.md:1]: 已修改

File: src/ffpa_attn/ffpa_attn_interface.py (hunk 4 above)
Comment (Butterfingrz) [reply src/ffpa_attn/ffpa_attn_interface.py]: 设置为显式指定 cutedsl 后端

File: src/ffpa_attn/ffpa_attn_interface.py (hunk 5 above)
Comment (Butterfingrz) [reply src/ffpa_attn/ffpa_attn_interface.py]: 全部移动到 src/ffpa_attn/cutedsl/_wrappers.py 中，清除了多余的代码逻辑

File: src/ffpa_attn/ffpa_attn_interface.py [hunk 7]
```diff
@@ -47,6 +85,16 @@ def _should_fallback_to_sdpa(
...
   assert D == D_k, "Query and key must have the same head dimension"
+  # CuTeDSL opt-in is explicit, so unsupported cases must raise rather
+  # than silently fall back to SDPA. cutedsl constraint checks split in
+  # two: tensor-level (head_dim == 512, SM90 device, dtype, training-bf16)
+  # are surfaced by _require_cutedsl_supported, and kwarg-level
+  # (dropout_p == 0, attn_mask is None, FA-extension kwargs) by
+  # _check_supported_options — both invoked from _ffpa_attn_cutedsl with
+  # actionable NotImplementedError / TypeError. Returning False here
+  # guarantees that path is reached.
+  if forward_backend == "cutedsl":
```
Comment (DefTruth) [inline src/ffpa_attn/ffpa_attn_interface.py:111]: 这里fallback到sdpa的逻辑还是需要，可能是前边的comment没说清楚，fwd/bwd backend之前暂不支持auto select, 但是fallback到sdpa的逻辑是需要的，它不是ffpa不同backend之间的auto select

**Review by DefTruth — CHANGES_REQUESTED** [review]

File: src/ffpa_attn/ffpa_attn_interface.py (hunk 7 above)
Comment (Butterfingrz) [reply src/ffpa_attn/ffpa_attn_interface.py:111]: head_dim != 512 或 非 SM90 时 fallback 到 SDPA
